Finally wire up the `sdk configure` settings to compilation

finagolfin opened the pull request:
This will require more discussion, but this attempts to fix #8958, will have more info soon.

Commit message: This only affects the native build system currently, with swift-build requiring more changes.

MaxDesiatov commented on Sources/PackageModel/SwiftSDKs/SwiftSDK.swift:
```diff
@@ -761,6 +761,50 @@ public struct SwiftSDK: Equatable {
+                observabilityScope.emit(
+                    warning: """
+                        Multiple SDKs have been configured with this triple \(swiftSDKSelector), so
+                            dropping back to values from original bundle.
```
I can have a look when this is ready for review and not marked as a work in progress draft.

MaxDesiatov replied:
(modulo any other work I and other people are doing, sometimes lead time for a review can be much longer than 2 days because of that)

MaxDesiatov replied:
In accordance with semantic versioning, these changes must be made in a backward-compatible manner with an appropriate deprecation period. I'm fine with these changes in principle, as long as we don't break anything for existing users, notify about the desired migration path, and give them time to migrate before the next major release.

finagolfin replied:
Alright, I will make the changes, let's see. One key question I asked a couple weeks ago still stands, can I substantially alter a [seemingly non-public extension method like `selectSwiftSDK(selector:hostTriple:observabilityScope)`](https://github.com/swiftlang/swift-package-manager/blob/4a6660ac95f4b2f88e1006f84576b0ca16343eb9/Sources/PackageModel/SwiftSDKs/SwiftSDKBundle.swift#L84) for both its behavior and declaration, or would I have to create a new method and deprecate the old one?

As for backwards compatibility and migration paths, the current behavior is buggy, non-deterministic, and incomplete for these two features, so pretty sure that making it strict and work right won't be "compatible" 😉, but we can try for the parts that were working.

finagolfin committed:
Fix copy-paste errors in using `sdk configure` to set the static resource path for Swift SDKs

finagolfin committed:
Finally wire up the `sdk configure` settings to compilation

This only affects the native build system currently, with swift-build requiring
more changes.

owenv reviewed (commented):
Thank you! I like this general direction, will give others a chance to review as well and make sure there aren't any high-level objections.

I do think it may be worth breaking out the `sdk configure` and Swift SDK selection changes into separate PRs, both to make them easier to review and because I think we may need to do a bit of work to get `sdk configure` working with the new build system backend (that is currently reading SDKs directly from disk and may need to handle the merging the in-memory customizations differently)

finagolfin commented on Sources/PackageModel/SwiftSDKs/SwiftSDKBundle.swift:
```diff
@@ -67,7 +67,7 @@ extension [SwiftSDKBundle] {
                     }
 
-                    return variant.swiftSDKs.first { $0.targetTriple == targetTriple }
+                    return variant.swiftSDKs.first { $0.targetTriple?.tripleString == targetTriple.tripleString }
```
Note that this `selectSwiftSDK(id: hostTriple: targetTriple)` "method was only used by the broken `swift sdk configure` before this pull," as I mentioned below, so it was essentially unused until I started using it for `--triple <triple>` checking in this pull and fixed `swift sdk configure` so it _can_ be used.

> I do wonder if we should allow a user-specified unversioned triple to match a versioned triple in the SDK if there is only one. Or potentially flag a 'near miss' as a descriptive error.

I'd rather not add these new features alongside these bug fixes. If you'd like those search features added, please file an issue and we can get them in later. This change merely gets `--triple <triple>` to work and [exactly matches the triple checking we've long had](https://github.com/swiftlang/swift-package-manager/pull/9229/changes#diff-c15b78e0314b27acd57daf9c85e41778ed1cb107b685bdaa6e7a552d4dce4791R103) with `--swift-sdk <triple>`, ie no new triple-matching behavior.

> your patch retains the "first" behavior... don't you need to change this to error if there's >1?

I thought about that too and kept it the way it was written, because I figured it is unlikely for SDK bundlers to make that mistake of specifying the same target triple multiple times and that the existing bundle verification probably already catches that issue.

However, I just tried manually editing the Android bundle to repeat some triples, and there is no such verification, either before or after installing: it just grabs the first matching triple it finds from multiple candidates in a bundle's JSON config. 🤨

The question now is how to address this unlikely potential problem in SDK bundles: should we fix it when [verifying the SDK before installing](https://github.com/swiftlang/swift-package-manager/blob/69f851d48eacdfc19c2cb94eb61e4c0858713474/Sources/PackageModel/SwiftSDKs/SwiftSDKBundleStore.swift#L317), as [we do for the name of the SDK itself](https://github.com/swiftlang/swift-package-manager/blob/69f851d48eacdfc19c2cb94eb61e4c0858713474/Sources/PackageModel/SwiftSDKs/SwiftSDKBundleStore.swift#L324), or put in some error-checking here when searching instead?

finagolfin commented:
Not ready yet, but just rebasing and checking linux CI

@swift-ci test linux

ken0nek commented on Sources/PackageModel/SwiftSDKs/SwiftSDK.swift:
```diff
@@ -786,7 +786,28 @@ public struct SwiftSDK: Equatable {
+                        targetTriple: swiftSDK.targetTriple!)
+                    {
+                        swiftSDK = configSDK
+                    }
```
Could `SwiftSDKConfigurationStore` grow a sibling method that returns just the persisted overlay (a `PathsConfiguration?`), so we can merge it into the SDK we already selected, instead of taking a re-selected full `SwiftSDK`? Roughly:

```swift
if let targetTriple = swiftSDK.targetTriple,
   let overlay = try configurationStore.readPathsOverlay(sdkID: id, targetTriple: targetTriple) {
    swiftSDK.pathsConfiguration.merge(with: overlay)
}
```

The existing `readConfiguration` would stay as-is for `swift sdk show` (`ConfigurationSubcommand.swift:67`) and the self-call in `configure()`. The new method would just sit alongside it for the compile path.

Reasoning:
1. `readConfiguration` re-walks `allValidBundles` and re-runs `selectSwiftSDK`, but we already have the bundle from `selectBundle` two lines up.
2. That re-selection passes `hostTriple: nil`. Per `SwiftSDKBundle.swift:56-76`, nil skips the host filter and returns the first variant's matching target, which for multi-host bundles can differ from the host-resolved variant we already have.
3. `readConfiguration` returns the SDK whether or not a configuration JSON exists, so `swiftSDK = configSDK` runs for every installed bundle, not just users who ran `swift sdk configure`. An overlay returning `nil` when there's no file gives a clean opt-in signal.
4. The configure subcommand stores path overrides, and `readConfiguration` merges only `pathsConfiguration` from the persisted JSON, so replacing the whole SDK reaches further than the configure data warrants.

finagolfin replied:
Good catch on 2., I don't think we have any multi-host SDKs yet, but that was a bug in waiting. I think we just tighten that up to a required parameter, as it is for the other `selectSwiftSDK(selector:hostTriple)` overload.

I don't think your efficiency concerns in 1. and 3. matter too much, as most people won't have many bundles installed, but 4. _is_ a good reason to implement what you suggest, as it minimizes what's modified, lessening the risk of bugs like 2.

I'll add that and update this pull.

finagolfin committed:
Always check the host triple in `selectSwiftSDK(id:hostTriple:targetTriple)`

finagolfin committed:
Remove deprecated `swift sdk configuration` subcommand

finagolfin committed:
Change `SwiftSDKConfigurationStore.readConfiguration()` to pass in a mutable SwiftSDK, rather than load all bundles every time its invoked

Refactor `SwiftSDKConfigurationStore.configure()` to use `selectSwiftSDK()` instead then
invoke this lighter `readConfiguration()`, and modify the other two calls of that method also.

finagolfin commented on Sources/PackageModel/SwiftSDKs/SwiftSDKBundle.swift:
```diff
@@ -53,18 +53,16 @@ extension [SwiftSDKBundle] {
-                            continue
-                        }
+                    guard variant.isSupporting(hostTriple: hostTriple) else {
+                        continue
```
Fixed the issue that @ken0nek found

finagolfin commented on Sources/PackageModel/SwiftSDKs/SwiftSDKConfigurationStore.swift:
```diff
@@ -85,56 +85,31 @@ public final class SwiftSDKConfigurationStore {
-    ) throws -> SwiftSDK? {
+        sdk: inout SwiftSDK
+    ) throws -> SwiftSDK {
+        let targetTriple = sdk.targetTriple!
```
@ken0nek, passing in the mutable SDK here let's me avoid loading all the bundles below, without having to pass the configuration around and merge it each time. Let me know what you think.

finagolfin commented on Sources/PackageModel/SwiftSDKs/SwiftSDKConfigurationStore.swift:
```diff
@@ -173,33 +148,34 @@ public final class SwiftSDKConfigurationStore {
-                    targetTriple: nil
-                )
-            }
+            sdks = bundles.selectSwiftSDK(matching: sdkID, hostTriple: hostTriple).idMatches
```
@marcprux, refactored this code to operate on SwiftSDKs and to use the existing `selectSwiftSDK()` methods, rather than your new private `swiftSDKs()` method that wasn't checking the host and would have hit [the same bug Ken raised above for future multi-host SDKs](https://github.com/swiftlang/swift-package-manager/pull/9229#discussion_r3149543976).
